cfn-resources/api-key/cmd/resource/resource.go

```diff
@@ -0,0 +1,543 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+
+	"github.com/aws-cloudformation/cloudformation-cli-go-plugin/cfn/handler"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/profile"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/constants"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/logger"
+	progress_events "github.com/mongodb/mongodbatlas-cloudformation-resources/util/progressevent"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/secrets"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/validator"
+	atlasSDK "go.mongodb.org/atlas-sdk/v20230201002/admin"
+)
+
+var CreateRequiredFields = []string{constants.OrgID, constants.Description}
+var UpdateRequiredFields = []string{constants.OrgID, constants.APIUserID, constants.Description}
+var ReadRequiredFields = []string{constants.OrgID, constants.APIUserID}
+var DeleteRequiredFields = []string{constants.OrgID, constants.APIUserID}
+var ListRequiredFields = []string{constants.OrgID}
+
+const (
+	CREATE = "CREATE"
+	READ   = "READ"
+	UPDATE = "UPDATE"
+	DELETE = "DELETE"
+	LIST   = "LIST"
+)
+
+type APIKeySecret struct {
+	PublicKey  string `json:"PublicKey"`
+	PrivateKey string
+}
+
+func setup() {
+	util.SetupLogger("mongodb-atlas-api-key")
+}
+
+// Create handles the Create event from the Cloudformation service.
+func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+
+	modelValidation := validator.ValidateModel(CreateRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
+
+	// Create atlas client
+	if currentModel.Profile == nil || *currentModel.Profile == "" {
+		currentModel.Profile = aws.String(profile.DefaultProfile)
+	}
+
+	atlas, peErr := util.NewAtlasClient(&req, currentModel.Profile)
+	if peErr != nil {
+		return *peErr, nil
+	}
+
+	// Set the roles from model
+	apiKeyInput := atlasSDK.CreateAtlasOrganizationApiKey{
+		Desc:  currentModel.Description,
+		Roles: currentModel.Roles,
+	}
+	apiKeyRequest := atlas.AtlasV2.ProgrammaticAPIKeysApi.CreateApiKey(
+		context.Background(),
+		*currentModel.OrgId,
+		&apiKeyInput,
+	)
+	apiKeyUserDetails, response, err := apiKeyRequest.Execute()
+	defer closeResponse(response)
+	if err != nil {
+		return handleError(response, CREATE, err)
+	}
+
+	// Read response
+	currentModel.APIUserId = apiKeyUserDetails.Id
+
+	// Save PrivateKey in AWS SecretManager
+	secret := APIKeySecret{PublicKey: *apiKeyUserDetails.PublicKey, PrivateKey: *apiKeyUserDetails.PrivateKey}
+	_, err = secrets.Create(&req, *currentModel.APIUserId, secret, currentModel.Description)
```

Comment:
here we are using the APIUserId to create the secret, it would be nice, if the user can provide the name, so we can use this resouce, to create the profile sercet, i have this idea on mind:
flow 1:
         the user creates a Stack:
- Cretes a Secret with the profile name (default for example)
- provides the secret name to this resource as an input and this resource stores the apikeys in that secret

flow 2:
the user provides the secret name, but the secret does not exists, so this resource will create the secret and store the apikeys on it

it would be nice to support both flows, but if we support the flow 1 it would be extreamly usefull on the Bootstrap

Reply:
a solution that i have on mind is that we can add a new filed that is AwsSecretName if the secret exists we add the apikeys on the secret, if it does not we create it

Reply:
AwsSecretName parameter added. If user doesn't provide The APIUserId is taken for secret name.